### PR TITLE
Introduce Test Framework Plugin Architecture

### PR DESCRIPTION
**Pull request by Maciek416**
This code is the implementation branch for:
  - https://github.com/TestArmada/magellan/issues/87

The code formerly known as `interop` has been moved to plugins:
  - https://github.com/TestArmada/magellan-mocha-plugin
  - https://github.com/TestArmada/magellan-nightwatch-plugin

Plugins are now installed and configured the following way:

```shell
npm install --save testarmada-magellan
npm install --save testarmada-magellan-mocha-plugin
```

`magellan.json`
```json
{
  "framework": "testarmada-magellan-mocha-plugin"
}
```

**Notes:**
  - Old framework names (`vanilla-mocha`, `nightwatch`, etc) are now translated to the official plugin names for those frameworks. If missing, magellan displays help on how to install.
  - `"nightwatch"` is still the default `framework` setting. As mentioned above, if it's missing, magellan outputs installation help.
  - A misleading setup/teardown message appearing on test suite failures has been fixed.
  - `TestRun` implementations no longer inherit from a base test run. They simply follow a signature (documentation to follow after PR).
  - Tests for framework support have been moved to their respective frameworks

### CODE REVIEW

File: bin/magellan [hunk 1]
```diff
@@ -55,6 +56,39 @@ if (margs.argv.list_browsers) {
...
+// NOTE: EXPERIMENTAL !! This will change before landing in master.
+//
+// NOTE: These should be required via settings/arguments. Note that
+//       magellan will NOT be including plugins directly, instead it will
+//       be necessary to include the plugin from a test suite project.
+//
+//
+//
+//
+if (settings.framework === "nightwatch") {
```
Comment (mattfysh) [inline bin/magellan]: could this be a magellan.json opt, opening up the way for other plugins?

Comment (Maciek416) [inline bin/magellan]: Yep, that's the idea.

**Comment by Maciek416** [comment]
@mattfysh take a look now. The name of the framework in `magellan.json` now **matches the name of the npm module**. We also translate the old framework name strings to new module names just for convenience (perhaps with a warning to migrate to proper module names later?)

**Comment by Maciek416** [comment]
I've run into a road slight block.

```
Homes-MacBook-Air:boilerplate-nightwatch home$ ./node_modules/.bin/magellan --create_tunnels
Loaded magellan configuration from:  /Users/home/projects/testarmada/boilerplate-nightwatch/magellan.json
Sauce configuration OK
Error: Could not start Magellan.
Error: Could not load the testing framework plugin 'testarmada-magellan-nightwatch-plugin'.
Check and make sure your package.json includes a module named 'testarmada-magellan-nightwatch-plugin'.
If it does not, you can remedy this by typing:

npm install --save testarmada-magellan-nightwatch-plugin
{ [Error: Cannot find module 'testarmada-magellan-nightwatch-plugin'] code: 'MODULE_NOT_FOUND' }
Homes-MacBook-Air:boilerplate-nightwatch home$ npm install --save testarmada-magellan-nightwatch-plugin
testarmada-boilerplate-nightwatch@1.1.1 /Users/home/projects/testarmada/boilerplate-nightwatch
├─┬ testarmada-magellan@5.0.0-beta -> /Users/home/projects/testarmada/magellan invalid
│ ├─┬ guacamole@1.1.4
│ │ └── cli-color@1.1.0 
│ └── lodash@3.10.1 
├─┬ testarmada-magellan-nightwatch@3.1.0-beta
│ └── lodash@3.10.1 
├─┬ testarmada-magellan-nightwatch-plugin@2.0.1 
│ ├── acorn@2.7.0 
│ ├── cli-color@1.1.0 
│ └── lodash@3.10.1 
└─┬ testarmada-performance-reporter@2.0.0
  └─┬ testarmada-magellan@3.0.1
    ├─┬ guacamole@1.1.4
    │ ├── cli-color@1.1.0 
    │ └── lodash@3.10.1 
    └── lodash@3.10.1 

Homes-MacBook-Air:boilerplate-nightwatch home$ ./node_modules/.bin/magellan --create_tunnels
Loaded magellan configuration from:  /Users/home/projects/testarmada/boilerplate-nightwatch/magellan.json
Sauce configuration OK
Error: Could not start Magellan.
Error: Could not load the testing framework plugin 'testarmada-magellan-nightwatch-plugin'.
Check and make sure your package.json includes a module named 'testarmada-magellan-nightwatch-plugin'.
If it does not, you can remedy this by typing:

npm install --save testarmada-magellan-nightwatch-plugin
{ [Error: Cannot find module 'testarmada-magellan-nightwatch-plugin'] code: 'MODULE_NOT_FOUND' }```

Looks like magellan can't see the plugin because it can't see your test suite's modules. Is `require("./node_modules/${PLUGIN_NAME}/index")` the right way to solve this?

**Comment by Maciek416** [comment]
Roadblock overcome.

**Comment by Maciek416** [comment]
This is now ready to rock

**Comment by geekdave** [comment]
:+1:

**Pull request merged**